File: contracts/wasm/erc20/go/erc20/erc20.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// implementation of ERC-20 smart contract for ISCP
// following https:// ethereum.org/en/developers/tutorials/understand-the-erc-20-token-smart-contract/

package erc20

import (
	"github.com/iotaledger/wasp/packages/vm/wasmlib/go/wasmlib"
)

// Sets the allowance value for delegated account
// inputs:
//  - PARAM_DELEGATION: agentID
//  - PARAM_AMOUNT: i64
func funcApprove(ctx wasmlib.ScFuncContext, f *ApproveContext) {
	delegation := f.Params.Delegation().Value()
	amount := f.Params.Amount().Value()
	ctx.Require(amount >= 0, "erc20.approve.fail: wrong 'amount' parameter")

	// all allowances are in the map under the name of he owner
	allowances := f.State.AllAllowances().GetAllowancesForAgent(ctx.Caller())
	allowances.GetInt64(delegation).SetValue(amount)
	f.Events.Approval(amount, ctx.Caller(), delegation)
}

// on_init is a constructor entry point. It initializes the smart contract with the
// initial value of the token supply and the owner of that supply
// - input:
//   -- PARAM_SUPPLY must be nonzero positive integer. Mandatory
//   -- PARAM_CREATOR is the AgentID where initial supply is placed. Mandatory
func funcInit(ctx wasmlib.ScFuncContext, f *InitContext) {
	supply := f.Params.Supply().Value()
	ctx.Require(supply > 0, "erc20.on_init.fail: wrong 'supply' parameter")
	f.State.Supply().SetValue(supply)

	// we cannot use 'caller' here because on_init is always called from the 'root'
	// so, owner of the initial supply must be provided as a parameter PARAM_CREATOR to constructor (on_init)
	// assign the whole supply to creator
	creator := f.Params.Creator().Value()
	f.State.Balances().GetInt64(creator).SetValue(supply)

	t := "erc20.on_init.success. Supply: " + f.Params.Supply().String() +
		", creator:" + creator.String()
	ctx.Log(t)
}

// transfer moves tokens from caller's account to target account
// This function emits the Transfer event.
// Input:
// - PARAM_ACCOUNT: agentID
// - PARAM_AMOUNT: i64
func funcTransfer(ctx wasmlib.ScFuncContext, f *TransferContext) {
	amount := f.Params.Amount().Value()
	ctx.Require(amount >= 0, "erc20.transfer.fail: wrong 'amount' parameter")

	balances := f.State.Balances()
	sourceAgent := ctx.Caller()
	sourceBalance := balances.GetInt64(sourceAgent)
	ctx.Require(sourceBalance.Value() >= amount, "erc20.transfer.fail: not enough funds")

	targetAgent := f.Params.Account().Value()
	targetBalance := balances.GetInt64(targetAgent)
	result := targetBalance.Value() + amount
	ctx.Require(result >= 0, "erc20.transfer.fail: overflow")

	sourceBalance.SetValue(sourceBalance.Value() - amount)
	targetBalance.SetValue(targetBalance.Value() + amount)

	f.Events.Transfer(amount, sourceAgent, targetAgent)
}

// Moves the amount of tokens from sender to recipient using the allowance mechanism.
// Amount is then deducted from the caller’s allowance.
// This function emits the Transfer event.
// Input:
// - PARAM_ACCOUNT: agentID   the spender
// - PARAM_RECIPIENT: agentID   the target
// - PARAM_AMOUNT: i64
func funcTransferFrom(ctx wasmlib.ScFuncContext, f *TransferFromContext) {
	// validate parameters
	amount := f.Params.Amount().Value()
	ctx.Require(amount >= 0, "erc20.transfer_from.fail: wrong 'amount' parameter")

	// allowances are in the map under the name of the account
	sourceAgent := f.Params.Account().Value()
	allowances := f.State.AllAllowances().GetAllowancesForAgent(sourceAgent)
	allowance := allowances.GetInt64(ctx.Caller())
	ctx.Require(allowance.Value() >= amount, "erc20.transfer_from.fail: not enough allowance")

	balances := f.State.Balances()
	sourceBalance := balances.GetInt64(sourceAgent)
	ctx.Require(sourceBalance.Value() >= amount, "erc20.transfer_from.fail: not enough funds")

	targetAgent := f.Params.Recipient().Value()
	targetBalance := balances.GetInt64(targetAgent)
	result := targetBalance.Value() + amount
	ctx.Require(result >= 0, "erc20.transfer_from.fail: overflow")

	sourceBalance.SetValue(sourceBalance.Value() - amount)
	targetBalance.SetValue(targetBalance.Value() + amount)
	allowance.SetValue(allowance.Value() - amount)

	f.Events.Transfer(amount, sourceAgent, targetAgent)
}

// the view returns max number of tokens the owner PARAM_ACCOUNT of the account
// allowed to retrieve to another party PARAM_DELEGATION
// Input:
// - PARAM_ACCOUNT: agentID
// - PARAM_DELEGATION: agentID
// Output:
// - PARAM_AMOUNT: i64
func viewAllowance(ctx wasmlib.ScViewContext, f *AllowanceContext) {
	// all allowances of the address 'owner' are stored in the map of the same name
	allowances := f.State.AllAllowances().GetAllowancesForAgent(f.Params.Account().Value())
	allow := allowances.GetInt64(f.Params.Delegation().Value()).Value()
	f.Results.Amount().SetValue(allow)
}

// the view returns balance of the token held in the account
// Input:
// - PARAM_ACCOUNT: agentID
func viewBalanceOf(ctx wasmlib.ScViewContext, f *BalanceOfContext) {
	balances := f.State.Balances()
	balance := balances.GetInt64(f.Params.Account().Value())
	f.Results.Amount().SetValue(balance.Value())
}

// the view returns total supply set when creating the contract (a constant).
// Output:
// - PARAM_SUPPLY: i64
func viewTotalSupply(ctx wasmlib.ScViewContext, f *TotalSupplyContext) {
	f.Results.Supply().SetValue(f.State.Supply().Value())
}

File: contracts/wasm/erc20/ts/erc20/erc20.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// implementation of ERC-20 smart contract for ISC
// following https://ethereum.org/en/developers/tutorials/understand-the-erc-20-token-smart-contract/

import * as wasmlib from "wasmlib"
import * as sc from "./index";

// Sets the allowance value for delegated account
// inputs:
//  - PARAM_DELEGATION: agentID
//  - PARAM_AMOUNT: u64
export function funcApprove(ctx: wasmlib.ScFuncContext, f: sc.ApproveContext): void {
    let delegation = f.params.delegation().value();
    let amount = f.params.amount().value();

    // all allowances are in the map under the name of he owner
    let allowances = f.state.allAllowances().getAllowancesForAgent(ctx.caller());
    allowances.getUint64(delegation).setValue(amount);
    f.events.approval(amount, ctx.caller(), delegation)
}

// onInit is a constructor entry point. It initializes the smart contract with the
// initial value of the token supply and the owner of that supply
// - input:
//   -- PARAM_SUPPLY must be nonzero positive integer. Mandatory
//   -- PARAM_CREATOR is the AgentID where initial supply is placed. Mandatory
export function funcInit(ctx: wasmlib.ScFuncContext, f: sc.InitContext): void {
    let supply = f.params.supply().value();
    ctx.require(supply > 0, "erc20.onInit.fail: wrong 'supply' parameter");
    f.state.supply().setValue(supply);

    // we cannot use 'caller' here because onInit is always called from the 'root'
    // so, owner of the initial supply must be provided as a parameter PARAM_CREATOR to constructor (onInit)
    // assign the whole supply to creator
    let creator = f.params.creator().value();
    f.state.balances().getUint64(creator).setValue(supply);

    let t = "erc20.onInit.success. Supply: " + supply.toString() +
        ", creator:" + creator.toString();
    ctx.log(t);
}

// transfer moves tokens from caller's account to target account
// This function emits the Transfer event.
// Input:
// - PARAM_ACCOUNT: agentID
// - PARAM_AMOUNT: u64
export function funcTransfer(ctx: wasmlib.ScFuncContext, f: sc.TransferContext): void {
    let amount = f.params.amount().value();

    let balances = f.state.balances();
    let sourceAgent = ctx.caller();
    let sourceBalance = balances.getUint64(sourceAgent);
    ctx.require(sourceBalance.value() >= amount, "erc20.transfer.fail: not enough funds");

    let targetAgent = f.params.account().value();
    let targetBalance = balances.getUint64(targetAgent);

    sourceBalance.setValue(sourceBalance.value() - amount);
    targetBalance.setValue(targetBalance.value() + amount);

    f.events.transfer(amount, sourceAgent, targetAgent)
}

// Moves the amount of tokens from sender to recipient using the allowance mechanism.
// Amount is then deducted from the caller’s allowance.
// This function emits the Transfer event.
// Input:
// - PARAM_ACCOUNT: agentID   the spender
// - PARAM_RECIPIENT: agentID   the target
// - PARAM_AMOUNT: u64
export function funcTransferFrom(ctx: wasmlib.ScFuncContext, f: sc.TransferFromContext): void {
    // validate parameters
    let amount = f.params.amount().value();

    // allowances are in the map under the name of the account
    let sourceAgent = f.params.account().value();
    let allowances = f.state.allAllowances().getAllowancesForAgent(sourceAgent);
    let allowance = allowances.getUint64(ctx.caller());
    ctx.require(allowance.value() >= amount, "erc20.transferFrom.fail: not enough allowance");

    let balances = f.state.balances();
    let sourceBalance = balances.getUint64(sourceAgent);
    ctx.require(sourceBalance.value() >= amount, "erc20.transferFrom.fail: not enough funds");

    let targetAgent = f.params.recipient().value();
    let recipientBalance = balances.getUint64(targetAgent);

    sourceBalance.setValue(sourceBalance.value() - amount);
    recipientBalance.setValue(recipientBalance.value() + amount);
    allowance.setValue(allowance.value() - amount);

    f.events.transfer(amount, sourceAgent, targetAgent)
}

// the view returns max number of tokens the owner PARAM_ACCOUNT of the account
// allowed to retrieve to another party PARAM_DELEGATION
// Input:
// - PARAM_ACCOUNT: agentID
// - PARAM_DELEGATION: agentID
// Output:
// - PARAM_AMOUNT: u64
export function viewAllowance(ctx: wasmlib.ScViewContext, f: sc.AllowanceContext): void {
    // all allowances of the address 'owner' are stored in the map of the same name
    let allowances = f.state.allAllowances().getAllowancesForAgent(f.params.account().value());
    let allow = allowances.getUint64(f.params.delegation().value()).value();
    f.results.amount().setValue(allow);
}

// the view returns balance of the token held in the account
// Input:
// - PARAM_ACCOUNT: agentID
export function viewBalanceOf(ctx: wasmlib.ScViewContext, f: sc.BalanceOfContext): void {
    let balances = f.state.balances();
    let balance = balances.getUint64(f.params.account().value());
    f.results.amount().setValue(balance.value());
}

// the view returns total supply set when creating the contract (a constant).
// Output:
// - PARAM_SUPPLY: u64
export function viewTotalSupply(ctx: wasmlib.ScViewContext, f: sc.TotalSupplyContext): void {
    f.results.supply().setValue(f.state.supply().value());
}
